src/dune/foreign_rules.ml

```diff
@@ -72,64 +72,54 @@ let include_dir_flags ~expander ~dir (stubs : Foreign.Stubs.t) =
+      match language with
+      | Foreign.Language.C -> Fdo.c_flags ctx
+      | Foreign.Language.Cxx -> Fdo.cxx_flags ctx
+    in
+    let flags = Foreign.Source.flags src in
+    Super_context.foreign_flags sctx ~dir ~expander ~flags ~language
+    |> Build.map ~f:(List.append ctx_flags)
   in
-  let c_flags = Build.map c_flags ~f:(List.append (Fdo.c_flags ctx)) in
-  Super_context.add_rule sctx ~loc
-    ~dir
-      (* With sandboxing we get errors like: bar.c:2:19: fatal error: foo.cxx:
-         No such file or directory #include "foo.cxx" *)
-    ~sandbox:Sandbox_config.no_sandboxing
-    (let src = Path.build (Foreign.Source.path src) in
-     Command.run
-     (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
-       (Ok ctx.ocamlc)
-       [ A "-g"
-       ; include_flags
-       ; Dyn (Build.map c_flags ~f:(fun x -> Command.quote_args "-ccopt" x))
-       ; A "-o"
-       ; Target dst
-       ; Dep src
-       ]);
-  dst
-
-let build_cxx_file ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
-  let flags = Foreign.Source.flags src in
-  let ctx = Super_context.context sctx in
   let output_param =
     match ctx.lib_config.ccomp_type with
     | Msvc -> [ Command.Args.Concat ("", [ A "/Fo"; Target dst ]) ]
     | Other _ -> [ A "-o"; Target dst ]
   in
-  let cxx_flags =
-    Super_context.foreign_flags sctx ~dir ~expander ~flags
-      ~language:Foreign.Language.Cxx
-  in
-  let cxx_flags = Build.map cxx_flags ~f:(List.append (Fdo.cxx_flags ctx)) in
   Super_context.add_rule sctx ~loc
     ~dir
-      (* this seems to work with sandboxing, but for symmetry with
-         [build_c_file] disabling that here too *)
+      (* With sandboxing we get errors like: bar.c:2:19: fatal error: foo.cxx:
+         No such file or directory #include "foo.cxx" (note: cxx compilation
+         does seems to work with sandboxing) *)
     ~sandbox:Sandbox_config.no_sandboxing
     (let src = Path.build (Foreign.Source.path src) in
      let c_compiler = Ocaml_config.c_compiler ctx.ocaml_config in
-     Command.run
      (* We have to execute the rule in the library directory as the .o is
-        produced in the current directory *) ~dir:(Path.build dir)
+        produced in the current directory *)
+     Command.run ~dir:(Path.build dir)
        (Super_context.resolve_program ~loc:None ~dir sctx c_compiler)
-       ( [ Command.Args.S [ A "-I"; Path ctx.stdlib_dir ]
-         ; include_flags
-         ; Command.Args.dyn cxx_flags
-         ]
-       @ output_param @ [ A "-c"; Dep src ] ));
+       (List.concat
+          Command.Args.
+            [ [ S [ A "-I"; Path ctx.stdlib_dir ]; include_flags; dyn flags ]
+            ; specific_flags
+            ; output_param
+            ; [ A "-c"; Dep src ]
+            ]));
   dst
 
+let build_c_file =
+  let specific_flags =
+    Command.Args.[ A "-fPIC"; A "-D_FILE_OFFSET_BITS=64"; A "-D_REENTRANT" ]
```

Comment:
`-fPIC` does not make sense on Windows. In general the flags added to the compiler by OCaml depend on the underlying system. You may want to retrieve them directly from `ocamlc -config`/`Makefile.config`.

Reply:
I am a bit confused about the various flag sets present in `ocamlc -config`: 
`ocamlc_cflags`, `ocamlopt_cflags`, `bytecomp_c_compiler` and `native_c_compiler`

What I don't understand is why `bytecomp_c_compiler` is not always equal to `gcc + ocamlc_cflags` (or `native_c_compiler` to gcc +  `ocamlopt_cflags`):
```
ocamlc_cflags: -O2 -fno-strict-aliasing -fwrapv -fPIC
bytecomp_c_compiler: gcc -O2 -fno-strict-aliasing -fwrapv -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT
``` 

On my Windows install it is the case for instance :
```
ocamlc_cflags: -march=x86-64 -mtune=generic -O2 -mms-bitfields
bytecomp_c_compiler: x86_64-w64-mingw32-gcc -march=x86-64 -mtune=generic -O2 -mms-bitfields
```


I am wondering if maybe `bytecomp_c_compiler` = `ocamlc_cflags` + `ocamlc_cppflags`. Would that make sense ?

Reply:
And the answer to my question is yes:

https://github.com/ocaml/ocaml/blob/8e15ad79df9e36fee3327ec0b8e3afaa51528153/utils/config.mlp#L49-L50

Reply:
The OCaml compiler doesn't know anything about C++. `cpp` refers to the C preprocessor.

Reply:
Shame on me ! 

That explains everything, thank you !